File: mobile/types/fragment.ts
/**
 * 碎片笔记类型定义
 */

import type { FragmentFolder } from './folder';

export type FragmentSource = 'voice' | 'manual' | 'video_parse';
export type FragmentAudioSource = 'upload' | 'external_link';

export interface SpeakerSegment {
  speaker_id: string;
  start_ms: number;
  end_ms: number;
  text: string;
}

export interface MediaAsset {
  id: string;
  media_kind: 'image' | 'audio' | 'file';
  original_filename: string;
  mime_type: string;
  file_size: number;
  checksum: string | null;
  width: number | null;
  height: number | null;
  duration_ms: number | null;
  status: string;
  created_at: string | null;
  file_url?: string | null;
  expires_at?: string | null;
}

export type LocalPendingImageUploadStatus =
  | 'pending'
  | 'uploading'
  | 'uploaded'
  | 'failed_pending_retry';

export interface LocalPendingImageAsset {
  local_asset_id: string;
  local_fragment_id: string;
  local_uri: string;
  mime_type: string;
  file_name: string;
  remote_asset_id?: string | null;
  upload_status: LocalPendingImageUploadStatus;
}

/**
 * 碎片笔记数据模型
 */
export interface Fragment {
  id: string;
  audio_file_url: string | null;
  audio_file_expires_at?: string | null;
  transcript: string | null;
  speaker_segments: SpeakerSegment[] | null;
  summary: string | null;
  tags: string[] | null;
  source: FragmentSource;
  audio_source?: FragmentAudioSource | null;
  created_at: string;
  updated_at: string;
  folder_id?: string | null;
  folder?: FragmentFolder | null;
  body_html: string;
  plain_text_snapshot?: string | null;
  content_state?: 'empty' | 'transcript_only' | 'body_present';
  media_assets?: MediaAsset[];
  audio_object_key?: string | null;
  backup_status?: 'pending' | 'synced' | 'failed';
  entity_version?: number;
  last_backup_at?: string | null;
  deleted_at?: string | null;
  is_filmed?: boolean;
  filmed_at?: string | null;
}

export interface FragmentListResponse {
  items: Fragment[];
  total: number;
  limit: number;
  offset: number;
}

export interface CreateFragmentRequest {
  transcript?: string;
  body_html?: string;
  summary?: string;
  tags?: string[];
  source?: FragmentSource;
  folder_id?: string;
  media_asset_ids?: string[];
}

export interface FragmentVisualizationPoint {
  id: string;
  x: number;
  y: number;
  z: number;
  transcript: string | null;
  summary: string | null;
  tags: string[] | null;
  source: FragmentSource;
  created_at: string | null;
  cluster_id: number | null;
  is_noise: boolean;
}

export interface FragmentVisualizationCentroid {
  x: number;
  y: number;
  z: number;
}

export interface FragmentVisualizationCluster {
  id: number;
  label: string;
  keywords: string[];
  fragment_count: number;
  centroid: FragmentVisualizationCentroid;
}

export interface FragmentVisualizationStats {
  total_fragments: number;
  clustered_fragments: number;
  uncategorized_fragments: number;
}

export interface FragmentVisualizationMeta {
  projection: string;
  clustering: string;
  used_vector_source: string;
}

export interface FragmentVisualizationResponse {
  points: FragmentVisualizationPoint[];
  clusters: FragmentVisualizationCluster[];
  stats: FragmentVisualizationStats;
  meta: FragmentVisualizationMeta;
}

File: backend/schemas/fragment.py
from typing import Optional

from pydantic import BaseModel, Field


class SpeakerSegmentItem(BaseModel):
    """说话人分段响应模型。"""

    speaker_id: str
    start_ms: int
    end_ms: int
    text: str


class FragmentCreate(BaseModel):
    """创建碎片笔记请求模型"""

    transcript: Optional[str] = Field(None, description="转写文本")
    source: str = Field("voice", description="来源：voice, manual, video_parse")
    audio_path: Optional[str] = Field(None, description="音频文件路径")

    class Config:
        json_schema_extra = {
            "example": {
                "transcript": "今天想到了一个关于定位的好点子",
                "source": "voice",
            }
        }


class FragmentItem(BaseModel):
    """碎片笔记响应模型"""

    id: str
    transcript: Optional[str]
    speaker_segments: Optional[list[SpeakerSegmentItem]] = None
    summary: Optional[str]
    tags: Optional[list[str]]
    source: str
    sync_status: str
    created_at: str
    audio_path: Optional[str] = None


class FragmentDetail(FragmentItem):
    """碎片笔记详情响应模型"""

    audio_path: Optional[str]


class FragmentListResponse(BaseModel):
    """碎片列表响应模型"""

    items: list[FragmentItem]
    total: int
    limit: int
    offset: int


class FragmentSimilarityQuery(BaseModel):
    """碎片语义查询请求模型"""

    query_text: str = Field(..., description="查询文本")
    top_k: int = Field(5, ge=1, le=20, description="返回结果数量")
    exclude_ids: list[str] = Field(default_factory=list, description="需要排除的碎片 ID")


class SimilarFragmentItem(FragmentItem):
    """相似碎片响应模型"""

    score: float
    metadata: dict = Field(default_factory=dict)


class FragmentVisualizationPoint(BaseModel):
    """碎片云图中的单个点位。"""

    id: str
    x: float
    y: float
    z: float
    transcript: Optional[str]
    summary: Optional[str]
    tags: Optional[list[str]]
    source: str
    sync_status: str
    created_at: Optional[str]
    cluster_id: Optional[int] = None
    is_noise: bool = False


class FragmentVisualizationCentroid(BaseModel):
    """聚类中心点坐标。"""

    x: float
    y: float
    z: float


class FragmentVisualizationCluster(BaseModel):
    """聚类信息。"""

    id: int
    label: str
    keywords: list[str] = Field(default_factory=list)
    fragment_count: int
    centroid: FragmentVisualizationCentroid


class FragmentVisualizationStats(BaseModel):
    """云图统计信息。"""

    total_fragments: int
    clustered_fragments: int
    uncategorized_fragments: int


class FragmentVisualizationMeta(BaseModel):
    """云图生成元数据。"""

    projection: str
    clustering: str
    used_vector_source: str


class FragmentVisualizationResponse(BaseModel):
    """碎片向量可视化响应。"""

    points: list[FragmentVisualizationPoint] = Field(default_factory=list)
    clusters: list[FragmentVisualizationCluster] = Field(default_factory=list)
    stats: FragmentVisualizationStats
    meta: FragmentVisualizationMeta
